### PR TITLE
functional tests

### PR DESCRIPTION
**Pull request by lukeapage**
Hi @sokra before I fully implement this, I wanted to check you'd like to take it, to save me effort if you disagree.

The integration tests at the moment just do snapshots of the transform.

My idea is to transform the tests using itself, then in the tests we can use jest-environment-dom and run the generated code.

The biggest question - should it replace what you have, or be an addition?

### CODE REVIEW

**Comment by lukeapage** [comment]
@sokra I went ahead anyway, would appreciate a review

I found some bugs, fixed them as I went along, but if it’s not manageable here I can split out

File: package.json [hunk 1]
```diff
@@ -22,6 +34,7 @@
     "@babel/cli": "^7.1.2",
     "@babel/preset-env": "^7.1.0",
     "@babel/preset-react": "^7.0.0",
+    "babel-core": "7.0.0-bridge.0",
```
Comment (sokra) [inline package.json]: Why include `@babel/core` and `babel-core`?

File: src/runtime/react/render.js [hunk 2]
```diff
@@ -1,14 +1,54 @@
...
 const map = new WeakMap();
 export default (element, parentNode) => {
 	let context = map.get(parentNode);
 	if (!context) {
 		map.set(parentNode, (context = { $$: { node: parentNode } }));
 	}
 	renderInternal(context, element, "node", false);
-	const node = context.node;
-	if (node.parentElement !== parentNode) parentNode.appendChild(node);
+	let node = context.node;
```
Comment (sokra) [inline src/runtime/react/render.js]: This duplicates a lot of logic from `src/runtime/replaceNode`. You should be able to reuse this logic.

File: src/runtime/replaceNode.js [hunk 3]
```diff
@@ -19,6 +19,9 @@ export default (oldNodes, newNodes) => {
 	oldNodes = toArray(oldNodes);
 	newNodes = toArray(newNodes);
 	const parentNode = oldNodes[0].parentNode;
+	if (!parentNode) {
```
Comment (sokra) [inline src/runtime/replaceNode.js]: This seem to be bug somewhere else. `replaceNode` should not be called on detached nodes.

File: test/__snapshots__/integration.test.js.snap [hunk 4]
```diff
@@ -1,43 +1,59 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (sokra) [inline test/__snapshots__/integration.test.js.snap]: I think these integration test doesn't make much sense. It doesn't really help finding bugs and reviewing the snapshot is nearly impossible to do correctly. Could you remove it? Your snapshot tests are far better.

File: test/basic.test.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import ReactDOM from 'react-dom';
+import * as Basic from './cases/basic';
```
Comment (sokra) [inline test/basic.test.js]: I feel that splitting these tests into two files is a bit disconnected. I think it's better to combine Components and tests into a single file.

File: test/basic.test.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+describe('basic', () => {
+	it('renders correctly', () => {
+		const { element, rendered } = renderHelper(<Basic.HelloWorld/>);
+
+		expect(rendered).toMatchSnapshot();
+		expect(rendered).toEqual(element.childNodes[0]);
+	});
+
+	it('renders updates from render', () => {
```
Comment (sokra) [inline test/basic.test.js]: I would be great if it's possible to run the test suite twice, once with react and once with rawact. This way the snapshots and test cases would automatically validate against native react.

File: test/cases/lifecycle.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	const returnedLog = log;
+	log = [];
+	return returnedLog;
+};
+
+export class Outer extends React.Component {
+	constructor() {
+		super();
+
+		this.UNSAFE_componentWillMount = logger('outer componentWillMount');
```
Comment (sokra) [inline test/cases/lifecycle.js]: I intentionally omitted these UNSAFE methods, since they are going to be removed anyway.

**Comment by lukeapage** [comment]
@sokra there is a prettier config, but its not a dependency. Do you plan to use it? if so I could use toMatchInlineSnapshot which I think would be nicer.

File: babel.config.js [hunk 8]
```diff
@@ -10,10 +10,22 @@ module.exports = {
...
+						spec: false,
+					},
+				],
 			]
+		},
+		{
+			include: ["./test", "./lib/runtime"],
+			presets: [
+					["@babel/preset-env", {
+						modules: 'cjs',
```
[the remark this answers is not in the document]
Comment (lukeapage) [reply babel.config.js]: ah, I didn't realize that the babel file will be picked up by a consumer. I guess it depends on their babel setup.
One option would also be to not publish this file, but the best option will be to move it under environment variables, which will also allow running with the real react.

File: src/runtime/react/render.js (hunk 2 above)
Comment (lukeapage) [reply src/runtime/react/render.js]: aha, yes. I just had to fix replaceNode to recognize NodeList as an array..

File: test/basic.test.js (hunk 6 above)
Comment (lukeapage) [reply test/basic.test.js]: great idea. done.

File: package.json (hunk 1 above)
Comment (lukeapage) [reply package.json]: its needed.
https://jestjs.io/docs/en/getting-started#using-babel

File: babel.config.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (lukeapage) [reply babel.config.js]: moved under babel env

**Comment by lukeapage** [comment]
@sokra changes applied. Almost every line has changed.. but its now much nice, thanks.
see my question above about prettier.

File: src/index.js [hunk 9]
```diff
@@ -127,14 +127,12 @@ export default declare(api => {
 						},
 						this
 					);
+					for (const global of this.globals.reverse())
```
Comment (lukeapage) [inline src/index.js]: fixes a problem with imports not able to be modified by other plugins, meaning output was a mix of cjs and es2015

File: src/numberToIdentifer.js [hunk 10]
```diff
@@ -2,7 +2,7 @@ const START_LOWERCASE_ALPHABET_CODE = "a".charCodeAt(0);
 const START_UPPERCASE_ALPHABET_CODE = "A".charCodeAt(0);
 const DELTA_A_TO_Z = "z".charCodeAt(0) - START_LOWERCASE_ALPHABET_CODE + 1;
 
-export default function numberToIdentifier(n)  {
+export default function numberToIdentifier(n) {
```
Comment (lukeapage) [inline src/numberToIdentifer.js]: I ran prettier on the source..

File: src/runtime/react/render.js [hunk 11]
```diff
@@ -1,14 +1,29 @@
...
+	if (Array.isArray(node)) {
+		if (parentNode.childNodes.length > 0) {
+			replaceNode(parentNode.childNodes, node);
+		} else {
```
Comment (lukeapage) [inline src/runtime/react/render.js]: alternative is making a more general purpose method used by replaceNode or by this.

**Comment by sokra** [comment]
Thanks

**Pull request merged**